packages/framework/attributor/src/runtimeAttributorDataStoreChannel.ts

```diff
@@ -0,0 +1,354 @@
+	MonitoringContext,
+	raiseConnectedEvent,
+	createChildMonitoringContext,
+} from "@fluidframework/telemetry-utils/internal";
+
+import { OpStreamAttributor, type IAttributor } from "./attributor.js";
+import { opBlobName, type IRuntimeAttributor } from "./attributorContracts.js";
+import { AttributorSerializer, chain, deltaEncoder, type Encoder } from "./encoders.js";
+import { makeLZ4Encoder } from "./lz4Encoder.js";
+
+/**
+ * Data store channel for the runtime attributor. This channel is responsible for storing and managing the
+ */
+export class RuntimeAttributorDataStoreChannel
+	extends TypedEventEmitter<IFluidDataStoreRuntimeEvents>
+	implements IFluidDataStoreChannel, IRuntimeAttributor
+{
+	public constructor(
+		public readonly dataStoreContext: IFluidDataStoreContext,
+		existing: boolean,
+	) {
+		super();
+		this.mc = createChildMonitoringContext({
+			logger: dataStoreContext.baseLogger,
+			namespace: "Attributor",
+		});
+		this._attachState = dataStoreContext.attachState;
+		if (existing) {
+			this.visibilityState =
+				dataStoreContext.attachState === AttachState.Detached
+					? VisibilityState.LocallyVisible
+					: VisibilityState.GloballyVisible;
+		} else {
+			this.visibilityState = VisibilityState.NotVisible;
+		}
+		// If it's existing we know it has been attached.
+		if (existing) {
+			this.deferredAttached.resolve();
+		}
+		this.entryPoint = new FluidObjectHandle<FluidObject>(
+			this,
+			"",
+			dataStoreContext.IFluidHandleContext,
+		);
+	}
+
+	public get IFluidDataStoreChannel(): IFluidDataStoreChannel {
+		return this;
+	}
+
+	public get IRuntimeAttributor(): IRuntimeAttributor {
+		return this;
+	}
+
+	public get(key: AttributionKey): AttributionInfo {
+		assert(
+			this.opAttributor !== undefined,
+			0x509 /* RuntimeAttributor must be initialized before getAttributionInfo can be called */,
+		);
+
+		if (key.type === "detached") {
+			throw new Error("Attribution of detached keys is not yet supported.");
+		}
+
+		if (key.type === "local") {
+			// Note: we can *almost* orchestrate this correctly with internal-only changes by looking up the current
+			// client id in the audience. However, for read->write client transition, the container might have not yet
+			// received a client id. This is left as a TODO as it might be more easily solved once the detached case
+			// is settled (e.g. if it's reasonable for the host to know the current user information at container
+			// creation time, we could just use that here as well).
+			throw new Error("Attribution of local keys is not yet supported.");
+		}
+
+		return this.opAttributor.getAttributionInfo(key.seq);
+	}
+
+	public has(key: AttributionKey): boolean {
+		if (key.type === "detached") {
+			return false;
+		}
+
+		if (key.type === "local") {
+			return false;
+		}
+
+		return this.opAttributor?.tryGetAttributionInfo(key.seq) !== undefined;
+	}
+
+	private encoder: Encoder<IAttributor, string> = {
+		encode: unreachableCase,
```

Comment:
Clever... this lets you avoid having to type it as maybe being `undefined`, even though it's not set until `initialize` is called?

Reply:
This was written by Abram, but I think the purpose you stated is correct.